Clean up temporary files created by forceToDiskExecution

ajohnson-stripe-zz opened the pull request:
This adds functionality for `WriteExecution` to keep track of temporary files created and then delete them after execution has finished. Currently this is only used by `forceToDiskExecution`, which previously would leave around temporary files every time it was used.

This fixes #1615.

ajohnson-stripe committed:
Clean up temporary files created by forceToDiskExecution

oscar-stripe commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -713,11 +742,11 @@ object Execution {
             (fn(conf, mode), otherFn(conf, mode))
           }
-          WriteExecution(head, h :: t ::: tail, newFn)
+          WriteExecution(head, h :: t ::: tail, newFn, tempFiles)
```
don't you need to merge the sets from the left and right? Seems like we are losing `this.tempFilesToCleanup`.

ajohnson-stripe-zz replied:
Ah good catch, I do need that here.

oscar-stripe commented on scalding-core/src/test/scala/com/twitter/scalding/ExecutionTest.scala:
```diff
@@ -69,6 +68,29 @@ class WordCountEc(args: Args) extends ExecutionJob[Unit](args) {
+      Set(tempFile)
+    }
+
+    Execution.write(writeFn, readFn, filesToDeleteFn)
```
can you test a `.zip` to be sure we don't lose files there?

oscar-stripe commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -378,7 +385,26 @@ object Execution {
+        try {
+          val path = new Path(file)
+          if (fs.exists(path)) {
+            fs.delete(path, true)
```
can you comment what the boolean means at this call site?

ajohnson-stripe-zz replied:
👍

oscar-stripe commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -378,7 +385,26 @@ object Execution {
+        case hdfsMode: HadoopMode =>  FileSystem.get(hdfsMode.jobConf)
+      }
+
+      filesToCleanup foreach { file: String =>
```
can we prefer the `.` style and use `filesToCleanup.foreach {`

ajohnson-stripe-zz replied:
👍

oscar-stripe reviewed (changes requested):
This is great! Thanks for attacking this problem.

ajohnson-stripe-zz commented:
Thanks for the  feedback @oscar-stripe. I incorporated in the changes you suggested.

I also noticed that the REPL expects that the temporary outputs from `forceToDiskExecution` stick around in things like the `snapshot` method on `ShellTypedPipe`. As such I moved the actual deletion to happen in a shutdown hook on JVM exit.

ajohnson-stripe-zz commented:
Hmm, I'm not sure I see why the CI build is failing. It's complaining about failing when compiling maple: `.[error] (maple/compile:compileIncremental) javac returned nonzero exit code`, but I only see warnings from the compiler, no errors. I can build that fine locally too.

I do see this in the log, maybe related?

```
.tar: Unexpected EOF in archive
tar: rmtlseek not stopped at a record boundary
tar: Error is not recoverable: exiting now
```

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -284,6 +288,34 @@ object Execution {
+          }
+        } catch {
+          // If we fail in deleting a temp file, log the error but don't fail the run
+          case e: Exception => LOG.info(s"Unable to delete temp file $file")
```
can we pass e to the log statement? that way we get the message and stacktrace? Also should this be a warn?

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -284,6 +288,34 @@ object Execution {
   /**
+    * This is a Thread used as a shutdown hook to clean up temporary files created by some Execution
+    */
+  case class TempFileCleanup(filesToCleanup: mutable.Set[String], mode: Mode) extends Thread {
```
since we're just iterating over the filesToCleanup in this case class, maybe we can make this a more generic interface rather than mutable.Set? Iterable / Seq / Set?

ajohnson-stripe-zz replied:
👍

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -284,6 +288,34 @@ object Execution {
+            // The "true" parameter here indicates that we should recursively delete everything under the given path
+            fs.delete(path, true)
+          }
+        } catch {
```
as this is part of a shutdown hook, should we add a catch all clause?

oscar-stripe replied:
Do you mean `catch { case e: Throwable =>` so we catch anything? Is that what you mean?

piyushnarang replied:
@oscar-stripe yeah. If we for whatever reason throw an caught exception in this code that will result in the shutdown code not being run completely as I understand it.

piyushnarang commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -284,6 +288,34 @@ object Execution {
   /**
+    * This is a Thread used as a shutdown hook to clean up temporary files created by some Execution
+    */
+  case class TempFileCleanup(filesToCleanup: mutable.Set[String], mode: Mode) extends Thread {
```
maybe also call out that if the job is aborted this will not cleanup stuff (if folks haven't read / know about shutdown hook semantics)

oscar-stripe replied:
+1 to taking an immutable `Iterable` here. Can we make the class private?

ajohnson-stripe-zz commented:
I've incorporated the next round of feedback. I'm still looking into it, but I'm not able to reproduce the CI build failure.

johnynek commented:
I restarted. Let's see if it is a CI issue.

On Tue, Nov 15, 2016 at 8:17 AM Andrew Johnson notifications@github.com
wrote:

> I've incorporated the next round of feedback. I'm still looking into it,
> but I'm not able to reproduce the CI build failure.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> https://github.com/twitter/scalding/pull/1621#issuecomment-260687072,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAEJdvQAZ-YG1YeGlmmZu_ynvCm-F0h0ks5q-dsLgaJpZM4Kxsnr
> .

ajohnson-stripe committed:
Try disabling travis caching

ajohnson-stripe committed:
Revert "Try disabling travis caching"

This reverts commit b9a04b6238c5cf8cea8ca939ddc7b65a13694678.

piyushnarang commented:
@ajohnson-stripe changes look good to me. 👍

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -313,7 +347,7 @@ object Execution {
         override protected[EvalCache] val messageQueue: LinkedBlockingQueue[EvalCache.FlowDefAction] = self.messageQueue
         override def start(): Unit = sys.error("Invalid to start child EvalCache")
-        override def finished(): Unit = sys.error("Invalid to finish child EvalCache")
+        override def finished(mode: Mode): Unit = sys.error("Invalid to finish child EvalCache")
```
don't we need to override the `addFilesToCleanup` to forward back to the parent list? I think any files added to the child will not be deleted, right?

ajohnson-stripe-zz replied:
Yeah, that does need to be added

ajohnson-stripe committed:
Forward filesToCleanup to parent in cleanCache

ajohnson-stripe committed:
consistency

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -305,15 +338,17 @@ object Execution {
     type Counters = Map[Long, ExecutionCounters]
     private[this] val cache = new FutureCache[(Config, Execution[Any]), (Any, Counters)]
     private[this] val toWriteCache = new FutureCache[(Config, ToWrite), Counters]
+    protected[EvalCache] val filesToCleanup = mutable.Set[String]()
```
actually, I think we should keep this private and override the add method. The reason is we forgot one more thing: this is not thread-safe, but we have so many threads in play, we should make this thread-safe.

I think we are modifying this value from several threads.

So, if we just forward the method, we can handle the locking a bit more cleanly inside that method.

ajohnson-stripe-zz replied:
Good point, I'll switch that up.

ajohnson-stripe committed:
synchronize in addFilesToCleanup

isnotinvain commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -284,6 +288,35 @@ object Execution {
+          }
+        } catch {
+          // If we fail in deleting a temp file, log the error but don't fail the run
+          case e: Throwable => LOG.warn(s"Unable to delete temp file $file", e)
```
style-wise I think `case e => LOG.warn(s"Unable to delete temp file $file", e)` is the same (without the `:Throwable` part)

ajohnson-stripe-zz replied:
It is the same, but leaving out the `: Throwable` generates a compiler warning requesting that you be explicit. I see a mix of both in the codebase now, but I'm happy to go with whatever's the preferred style.

isnotinvain commented on scalding-core/src/main/scala/com/twitter/scalding/Execution.scala:
```diff
@@ -284,6 +288,35 @@ object Execution {
+    *
+    * If the job is aborted the shutdown hook may not run and the temporary files will not get cleaned up
+    */
+  private[scalding] case class TempFileCleanup(filesToCleanup: Iterable[String], mode: Mode) extends Thread {
```
should this extend Runnable instead? Does it have to extend Thread?

ajohnson-stripe-zz replied:
`addShutdownHook` will only take a `Thread` unfortunately

johnynek commented:
👍 looks good to me.

johnynek commented:
For throwable we should be explicit and avoid the warning in my view.
On Tue, Nov 15, 2016 at 12:05 Andrew Johnson notifications@github.com
wrote:

> ## _@ajohnson-stripe_ commented on this pull request.
> 
> In scalding-core/src/main/scala/com/twitter/scalding/Execution.scala
> https://github.com/twitter/scalding/pull/1621:
> 
> > -    override def run(): Unit = {
> > -      val fs = mode match {
> > -        case localMode: CascadingLocal => FileSystem.getLocal(new Configuration)
> > -        case hdfsMode: HadoopMode =>  FileSystem.get(hdfsMode.jobConf)
> > -      }
> >   +
> > -      filesToCleanup.foreach { file: String =>
> > -        try {
> > -          val path = new Path(file)
> > -          if (fs.exists(path)) {
> > -            // The "true" parameter here indicates that we should recursively delete everything under the given path
> > -            fs.delete(path, true)
> > -          }
> > -        } catch {
> > -          // If we fail in deleting a temp file, log the error but don't fail the run
> > -          case e: Throwable => LOG.warn(s"Unable to delete temp file $file", e)
> 
> It is the same, but leaving out the : Throwable generates a compiler
> warning requesting that you be explicit. I see a mix of both in the
> codebase now, but I'm happy to go with whatever's the preferred style.
> 
> —
> You are receiving this because you commented.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/twitter/scalding/pull/1621, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAEJdpRzfsyQs4txuM5bz7H4UcdXfz_9ks5q-hCHgaJpZM4Kxsnr
> .
